strawberry/fastapi/router.py

```diff
@@ -30,10 +31,30 @@ async def __get_root_value():
         return None
 
     @staticmethod
-    async def __get_context(
-        background_tasks: BackgroundTasks, request: Request = None, ws: WebSocket = None
-    ) -> Optional[Any]:
-        return {"request": request or ws, "background_tasks": background_tasks}
+    def __get_context_getter(
+        custom_getter: Callable[..., Optional[Dict[str, Any]]]
+    ) -> Callable[..., Dict[str, Any]]:
+        async def dependency(
+            background_tasks: BackgroundTasks,
+            request: Request = None,
+            ws: WebSocket = None,
+            custom_getter: Dict[str, Any] = {},
+        ):
+            return {
+                "request": request or ws,
+                "background_tasks": background_tasks,
+                **custom_getter,
+            }
```

Comment:
It really doesn't matter if it's a positional parameter or _keyworded_ one, as you'll modify the signature below.

The [mutable default value](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html) is unnecessary on the dependency.

```suggestion
        async def dependency(
            custom_getter: Dict[str, Any],
            background_tasks: BackgroundTasks,
            request: Request = None,
            ws: WebSocket = None,
        ):
            return {
                "request": request or ws,
                "background_tasks": background_tasks,
                **custom_getter,
            }
```
The order also doesn't matter, jfyk.

Reply:
Wow, I never knew about that dictionary anti-pattern. Thanks for informing me!

Reply:
Be aware that your code would have never reached the antipattern behavior. I'm just mentioning because it is not really necessary, and it avoids future PRs about "fixing" what is actually not doing anything :kissing:

Reply:
This change causes: `msg        = "duplicate parameter name: 'custom_getter'"` due to `parameters = [<Parameter "background_tasks: starlette.background.BackgroundTasks">, <Parameter "custom_getter: Dict[str, Any]">, <P...eter "request: starlette.requests.Request = None">, <Parameter "custom_getter: Dict[str, Any] = Depends(get_context)">]`, so a default value is required. I'll set it to `None` though to avoid that anti-pattern.

Reply:
There were errors caused by this: `parameters = [<Parameter "background_tasks: starlette.background.BackgroundTasks">, <Parameter "custom_getter: Dict[str, Any]">, <P...eter "request: starlette.requests.Request = None">, <Parameter "custom_getter: Dict[str, Any] = Depends(get_context)">]`, so line 57 of `router.py` should read
```python
sig.parameters["custom_getter: Dict[str, Any]"].replace(default=Depends(custom_getter))`.
```

Is this unusual that the type signatures are being supplied to `signature()` by `dependency()`?

Reply:
ups! I'm sorry, the order matters because we're pasting the n - 1 parameters back to the signature :sweat_smile: 

![image](https://user-images.githubusercontent.com/7353520/145865455-7cf61a03-b797-41a2-8725-9d6c740963a0.png)

Reply:
Aha, that's what that cryptic line is doing! Your method makes sense to me now, thanks for the help :)